cpp/include/cudf/detail/gather.cuh

```diff
@@ -46,58 +46,82 @@ struct bounds_checker {
+                                      size_type mask_num_bits,
+                                      size_type* validity_counts)
+{
+  for (size_type mask_idx = 0; mask_idx < mask_count; mask_idx++) {
+    auto const mask = masks[mask_idx];
+    if (mask == nullptr) {
+      continue;
+    }
 
+    auto bit_idx_offset = blockIdx.x * blockDim.x;
+    auto validity_count_aggregate = static_cast<size_type>(0);
 
-template <bool ignore_out_of_bounds, typename MapIterator>
-__global__ void gather_bitmask_kernel(table_device_view source_table,
-                                      MapIterator gather_map,
-                                      bitmask_type * masks[],
-                                      size_type destination_table_num_rows,
-                                      size_type* valid_counts) {
-
-  for (size_type i = 0; i < source_table.num_columns(); i++) {
-
-    column_device_view source_col = source_table.column(i);
-
-    if (masks[i] != nullptr) {
-      size_type destination_row_base = blockIdx.x * blockDim.x;
-      cudf::size_type valid_count_accumulate = 0;
-
-      while (destination_row_base < destination_table_num_rows) {
-        size_type destination_row = destination_row_base + threadIdx.x;
-
-        const bool thread_active = destination_row < destination_table_num_rows;
-        size_type source_row =
-          thread_active ? gather_map[destination_row] : 0;
-
-        bool bit_is_valid;
-        if (ignore_out_of_bounds && (source_row < 0 || source_row >= source_col.size())) {
-          bit_is_valid = thread_active && bit_is_set(masks[i], destination_row);
-        } else {
-          bit_is_valid = source_col.is_valid(source_row);
-        }
-
-        // Use ballot to find all valid bits in this warp and create the output
-        // bitmask element
-        const uint32_t valid_warp =
-          __ballot_sync(0xffffffff, thread_active && bit_is_valid);
-
-        const size_type valid_index = word_index(destination_row);
-
-        // Only one thread writes output
-        if (0 == threadIdx.x % warp_size and thread_active) {
-          masks[i][valid_index] = valid_warp;
-        }
-        valid_count_accumulate += single_lane_block_popc_reduce(valid_warp);
-        destination_row_base += blockDim.x * gridDim.x;
-      }
-      if (threadIdx.x == 0) {
-        atomicAdd(valid_counts + i, valid_count_accumulate);
+    while (bit_idx_offset < mask_num_bits) {
+      auto const bit_idx = bit_idx_offset + threadIdx.x;
+      auto const bit_active = bit_idx < mask_num_bits;
+      auto const bit_is_valid = bit_active && get_validity(mask_idx, bit_idx);
+      auto const warp_validity = __ballot_sync(0xffffffff, bit_is_valid);
+      auto const validity_idx = word_index(bit_idx);
+
+      if (bit_active && threadIdx.x % warp_size == 0) {
+        mask[validity_idx] = warp_validity;
       }
+
+      validity_count_aggregate += single_lane_block_popc_reduce(warp_validity);
```

Comment:
I didn't notice before, but this is the older version of this utility. Here is the newer, more general version which should be used instead: https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/include/cudf/detail/utilities/cuda.cuh#L87